plugin/clone/src/clone_common.cc

```diff
@@ -0,0 +1,111 @@
+/*
+  Copyright (C) 2022, Laurynas Biveinis
+
+  This program is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; version 2 of the License.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA */
+
+#include "plugin/clone/include/clone_common.h"
+
+#include <string>
+#include <utility>
+
+#include "lex_string.h"
+#include "my_dbug.h"
+#include "my_inttypes.h"
+#include "plugin/clone/include/clone_hton.h"
+#include "sql/handler.h"
+#include "sql/json_dom.h"
+#include "sql/sql_plugin_ref.h"
+#include "storage/perfschema/table_log_status.h"
+
+namespace myclone {
+
+// First, call precopy for the participating storage engines that have the
+// precopy method. Then, for master thread only, perform the cross-engine
+// synchronization: execute a performance_schea.log_status query, and call
+// set_log_stop for each storage engine with its part of STORAGE_ENGINES column
+// JSON object from that query.
+int Ha_clone_common_cbk::precopy_and_synchronize_engines(THD *thd,
+                                                         uint task_id) {
+  const auto &all_locators = get_all_locators();
+
+  // Must be called by InnoDB the clone-driving engine
+  assert(all_locators[0].m_hton->db_type == DB_TYPE_INNODB);
+  assert(get_loc_index() == 0);
+  auto *const saved_cbk_hton = get_hton();
+
+  uint index = 0;
+  for (const auto &loc : all_locators) {
+    auto *const precopy_fn = loc.m_hton->clone_interface.clone_precopy;
+    if (precopy_fn) {
+      set_loc_index(index);
+      const auto res =
+          precopy_fn(loc.m_hton, thd, loc.m_loc, loc.m_loc_len, task_id, this);
+      if (res != 0) {
+        set_loc_index(0);
+        set_hton(saved_cbk_hton);
+        return res;
+      }
+    }
+    ++index;
+  }
+  set_loc_index(0);
+  set_hton(saved_cbk_hton);
+
+  if (task_id != 0) return 0;
+
+  std::unique_ptr<table_log_status> table{static_cast<table_log_status *>(
+      table_log_status::create(&table_log_status::m_share))};
+
+  auto err [[maybe_unused]] = table->rnd_init(true);
+  assert(err == 0);
+
+  err = table->rnd_next();
+  assert(err == 0);
+
+  auto &log_status_row = table->get_row();
+  const auto &se_positions = log_status_row.w_storage_engines;
+  assert(se_positions.is_dom());
+
+  const auto *const json_dom = se_positions.get_dom();
+  assert(json_dom->json_type() == enum_json_type::J_OBJECT);
+
+  const auto *const json_obj = static_cast<const Json_object *>(json_dom);
```

Comment:
Regarding the memory leakage issue we met in the precopy_and_synchronize_engines function. According to the stack trace error message, the last place we hit the error is from stl_tree.h, which is used to implement associative containers like map, set. Looks like the json_dom/json_object is the only object using map. I am thinking aloud. When downcasting via static_cast, since the static_cast itself does not provide any safety check, do we need to do any safety check to avoid some dereference issue later?

Reply:
@sunshine-Chun, that is one option after we figure out why the JSON document here is not of the expected shape in the first place. So far I could not find any circumstances why it would ever be

Reply:
So the release build never checked for an error because I thought that it is impossible for any error to occur here. It seems we'll have to check and abort, but hopefully we'll still learn in what circumstances what kind of errors may happen